### ISSUE TITLE
Support DTS in WAV

### ISSUE THREAD

**Issue by shenbuqingyun** [google/ExoPlayer#9096]
Hello there, I have add ffmpeg extension, and --enable-decoder=ac3, --enable-decoder=dca.

And ExoPlayer does play AC3 correctly, but not a dts.

a wav song, with ffprobe  -show_format&-show_streams  info:
 "codec_name": "dts"
 "codec_long_name": "DCA (DTS Coherent Acoustics)"
 "profile": "DTS-ES"
 "channels": 7
 "filename": "new.wav"
 "format_name": "wav"
"format_long_name": "WAV / WAVE (Waveform Audio)"

it`s voice is "sasasa", when I play it.

EXO FfmpegLibrary will return abdio/raw, not audio/dts.
so it`s voice is incorrect.

You can download a wav song with it`s codec fomat is DTS, not pcm_s16le, give it a try.

And by the way, I have setExtensionRendererMode(EXTENSION_RENDERER_MODE_PREFER).

My Exo is 2.14.1, maybe it has no concern with this.

**Comment by shenbuqingyun** [comment]
https://www.dtshot.com/category/DTS/

it is a website, with dts song.

**Comment by shenbuqingyun** [comment]
when handle a dts wav song, FfmpegLibrary.java - getCodecName() will not get codec name audio/dts, but audio/raw, so getCodecName() will return null, it means will use MediaCodecAudioRenderer, not FfmpegAudioRenderer to handle this song.

it`s voice indeed is "sasasa".

If I change FfmpegLibrary.java - getCodecName() to return audio/dca compulsorily, ExoPlayer will play this song with error "AudioTrack: Discontinuity detected", like this:
29086-29167 E/AudioTrack: Discontinuity detected [expected 29169660, got 29400000]
29086-29167 E/AudioTrack: Discontinuity detected [expected 29469660, got 29700000]
29086-29167 E/AudioTrack: Discontinuity detected [expected 29769660, got 30000000]

**Comment by icbaker** [comment]
@kim-vde I think the root of this question is about `WavExtractor` supporting more than just PCM data.

**Comment by shenbuqingyun** [comment]
Yes, indeed.

for wav song， it`s default decoder name is pcm_s16le，but there are some other decoders like dts, WavExtractor can not read them correctly. 

for a ac3, Exo have a Ac3Extractor to handle it, after add ffmpeg extension, ac3 will play correctly.

but not dts, There is not a Extractor for it.

currently, I can use ffmpg library to transfer a local dts song to ac3, and use ExoPlayer to play it. But this way is useless for a net song.

**Comment by shenbuqingyun** [comment]
> @kim-vde I think the root of this question is about `WavExtractor` supporting more than just PCM data.

Yes, indeed.

for a wav song， it`s default decoder is pcm_s16le，but there are some other decoders like dts, WavExtractor can not read them correctly. 

for a ac3, Exo have a Ac3Extractor to handle it, after add ffmpeg extension, ac3 will play correctly.

but not dts, There is not a Extractor for it.

currently, I can use ffmpg library to transfer a local dts song to ac3, and use ExoPlayer to play it. But this way is useless for a net song. 

What should I do to solve this problm?

**Comment by shenbuqingyun** [comment]
Hello there,  again

Exoplayer may have a plan about to improve the relevant support of wav dts?

or is there anything I can do？ 

Thank you very much.

**Comment by kim-vde** [comment]
Indeed, the `WavExtractor` only supports IMA ADPCM, A-law, Mu-law and raw audio at the moment.

Could you please provide a test content with DTS + WAV (I can't download them from the website you provided)? If you're unable to share the test content publicly, please send it to dev.exoplayer@gmail.com using a subject "Issue #9096". Please also update this issue to indicate you've done this.

**Comment by moneytoo** [comment]
Some publicly accessible DTS WAV samples can be found here:

* http://samples.mplayerhq.hu/A-codecs/DTS/03_NEUE_MUSIK_BERND_THEWESl_kleine_pingpark_wolke_DTS.wav
* http://samples.mplayerhq.hu/A-codecs/DTS/06_NEW_JAZZ_SSQ3_ENERGY_DTS.wav
* http://samples.mplayerhq.hu/A-codecs/DTS/Fire__VooDoo_Studio_30_sec.wav
* http://samples.mplayerhq.hu/A-codecs/DTS/dts/dtswavsample14.wav

**Comment by kim-vde** [comment]
We use the header format tag to determine the wav format ([code](https://github.com/google/ExoPlayer/blob/r2.14.2/library/extractor/src/main/java/com/google/android/exoplayer2/extractor/wav/WavExtractor.java#L98)). I only checked the first sample provided in the comment above but it has format tag 0x0001, which normally corresponds to PCM. There is probably another way to determine that it is DTS audio, because the file plays with ffplay. 

Feel free to make a pull request to support wav + dts and we will review it. In the mean time, I will mark this issue as a low priority enhancement.

**Comment by shenbuqingyun** [comment]
Sorry, I did not receive the notification from github, I missed it. 

I send a email to dev.exoplayer@gmail.com which contains a dts-wav song, please check.

Sorry again.

**Comment by kim-vde** [comment]
Are you sure you sent the email? I can't find any email with subject "Issue #9096".

**Comment by kim-vde** [comment]
Thanks for providing the sample. The format tag is also 0x0001 for the file you provided, so the conclusion is the same as in my previous comment.

**Comment by icbaker** [comment]
Closing as a duplicate of https://github.com/androidx/media/issues/2015 to move the issue to the media3 issue tracker.
